src/ristretto/dalek_range_proof.rs

```diff
@@ -88,6 +90,60 @@ impl RangeProofService for DalekRangeProofService {
     fn range(&self) -> usize {
         self.range
     }
+
+    fn construct_proof_with_rewind_key(&self, key: &RistrettoSecretKey, value: u64, rewind_key: &RistrettoSecretKey, rewind_blinding_key: &RistrettoSecretKey, proof_message: &[u8; 19]) -> Result<Vec<u8>, RangeProofError> {
+        let mut pt = Transcript::new(b"tari");
+        let mut full_proof_message = [0u8; 23];
+        full_proof_message[0..4].clone_from_slice(b"tari");
+        full_proof_message[4..].clone_from_slice(proof_message);
+
+        let k = key.0;
+        let rk = rewind_key.0;
+        let rbk = rewind_blinding_key.0;
+        let (proof, _) = DalekProof::prove_single_with_rewind_key(&self.bp_gens, &self.pc_gens, &mut pt, value, &k, self.range, &rk, &rbk, &full_proof_message)
+            .map_err(|_| RangeProofError::ProofConstructionError)?;
+        Ok(proof.to_bytes())
+    }
+
+    fn rewind_proof_value_only(&self, proof: &Self::P, commitment: &PedersenCommitment, rewind_public_key: &RistrettoPublicKey, rewind_blinding_public_key: &RistrettoPublicKey) -> Result<(u64, [u8; 19]), RangeProofError> {
+        let rp = DalekProof::from_bytes(&proof).map_err(|_| RangeProofError::InvalidProof);
+        if rp.is_err() {
+            return Err(RangeProofError::InvalidRangeProof);
+        }
+        let rp = rp.unwrap();
+        let mut pt = Transcript::new(b"tari");
+        let rewind_nonce_1 = get_rewind_nonce_from_pub_key(&rewind_public_key.compressed, &commitment.as_public_key().compressed);
+        let rewind_nonce_2 = get_rewind_nonce_from_pub_key(&rewind_blinding_public_key.compressed, &commitment.as_public_key().compressed);
+        let (confidential_value, proof_message) = rp.rewind_single_get_value_only(&self.bp_gens, &mut pt, &commitment.as_public_key().compressed, self.range,&rewind_nonce_1,&rewind_nonce_2).map_err(|_| RangeProofError::ProofConstructionError)?;
+        if &proof_message[..4] != b"tari" {
+            return Err(RangeProofError::InvalidRewind);
+        }
+        let mut truncated_proof_message: [u8; 19] = [0u8; 19];
+        truncated_proof_message.copy_from_slice(&proof_message[4..]);
+        Ok((confidential_value, truncated_proof_message))
+    }
+
+    fn rewind_proof_commitment_data(&self, proof: &Self::P, commitment: &PedersenCommitment, rewind_key: &RistrettoSecretKey, rewind_blinding_key: &RistrettoSecretKey) -> Result<(u64, RistrettoSecretKey, [u8; 19]), RangeProofError> {
+        let rp = DalekProof::from_bytes(&proof).map_err(|_| RangeProofError::InvalidProof);
+        if rp.is_err() {
+            return Err(RangeProofError::InvalidRangeProof);
+        }
+        let rp = rp.unwrap();
```

Comment:
Same redundant err map

Reply:
agreed, copy-pasta from the existing verify code and didn't really think about it, will improve it there too.